streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -908,20 +908,25 @@ private void initializeAndRestorePhase() {
 
     // Check if the topology has been updated since we last checked, ie via #addNamedTopology or #removeNamedTopology
     private void checkForTopologyUpdates() {
-        if (lastSeenTopologyVersion < topologyMetadata.topologyVersion() || topologyMetadata.isEmpty()) {
-            lastSeenTopologyVersion = topologyMetadata.topologyVersion();
-            taskManager.handleTopologyUpdates();
-
+        do {
             topologyMetadata.maybeWaitForNonEmptyTopology(() -> state);
+            if (lastSeenTopologyVersion < topologyMetadata.topologyVersion()) {
```

Comment:
I feel this is getting more complicated than necessary to do KAFKA-12648: we have a `topologyMetadata` at the `KafkaStreams` layer while at the `StreamThread` layer we keep a `long lastSeenTopologyVersion`. If we let the `topologyMetadata` object which is shared among all threads (as well as their task managers etc) then we can simplify this. Just a sketchy thought here:

* In `TopologyMetadata`  we maintain a `Map<String, Long>` from thread name to thread's current topology version. When new threads are added / threads are removed, this map would be updated as well (in synchronized way).
* Inside a thread's `checkForTopologyUpdate`, in a synchronized block check if all threads' versions except this thread is equal to the current version: if yes, this thread would update its corresponding map entry as well and then trigger rebalance; otherwise, only update the corresponding map.

So if consecutive topology updates are being issued, then we would only trigger a rebalance at the end when all threads reaches the end topology version.

Reply:
As for the future: I think it's okay to let the first future be waiting a bit more if there are consecutive operations, i.e. in the above implementation we would complete all futures that have been registered so far when we eventually get every still-alive threads to catch up with the bumped version.

Reply:
Yeah I'm not quite sure these changes are necessary, I thought I had it set up to behave the way we want/need in the last PR I just merged. Of course I could have missed something that had to be added for this PR, but can you explain a bit more why we need these changes? Was this related to trying to fix tests that had broken?

Reply:
The previous version was not working because when the last topology was removed it would ungate and proceed  to the poll phase and throw an exception. Let me see what I can simplify using @guozhangwang's suggestions and then you can give it another look